### PR TITLE
fix: change top tokens query return type

### PR DESCRIPTION
**Pull request by kayleegeorge**
change top tokens query return type from record to TokenData[], fixes favoriting reload behavior as well

### CODE REVIEW

File: src/components/Tokens/TokenTable/TokenRow.tsx [hunk 1]
```diff
@@ -432,10 +432,9 @@ export default function LoadedRow({
   tokenData: TokenData
   timePeriod: TimePeriod
 }) {
-  const token = useToken(tokenAddress)
   const currency = useCurrency(tokenAddress)
-  const tokenName = token?.name ?? ''
-  const tokenSymbol = token?.symbol ?? ''
+  const tokenName = tokenData.name ?? ''
```
Comment (vm) [inline src/components/Tokens/TokenTable/TokenRow.tsx:436]: i don't actually like defaulting to empty string here. why is it necessary?

File: src/components/Tokens/TokenTable/TokenTable.tsx [hunk 2]
```diff
@@ -48,34 +47,29 @@ const TokenRowsContainer = styled.div`
   width: 100%;
 `
 
-function useFilteredTokens(addresses: string[]) {
+function useFilteredTokens(tokens: TokenData[]) {
```
Comment (vm) [inline src/components/Tokens/TokenTable/TokenTable.tsx:50]: should `useFilteredTokens` and `useSortedTokens` be conslidated?

File: src/graphql/data/TopTokenQuery.ts [hunk 3]
```diff
@@ -1,8 +1,37 @@
...
+  YEAR,
+  ALL,
+}
+
+interface IAmount {
+  currency: Currency | null
+  value: number | null
+}
+
+export type TokenData = {
```
Comment (vm) [inline src/graphql/data/TopTokenQuery.ts:20]: how could it be null or undefined?

File: src/pages/Tokens/index.tsx [hunk 4]
```diff
@@ -91,7 +91,7 @@ const Tokens = () => {
         </FiltersWrapper>
 
         <TokenTableContainer>
-          <TokenTable data={topTokens} error={error} loading={loading} />
+          <TokenTable data={topTokens ?? []} error={error} loading={loading} />
```
Comment (vm) [inline src/pages/Tokens/index.tsx:94]: better to allow TokenTable to take undefined for data imo

File: src/components/Tokens/TokenTable/TokenRow.tsx [hunk 5]
```diff
@@ -432,10 +432,9 @@ export default function LoadedRow({
   tokenData: TokenData
   timePeriod: TimePeriod
 }) {
-  const token = useToken(tokenAddress)
   const currency = useCurrency(tokenAddress)
-  const tokenName = token?.name ?? ''
-  const tokenSymbol = token?.symbol ?? ''
+  const tokenName = tokenData.name
+  const tokenSymbol = tokenData.symbol?.toUpperCase()
```
Comment (vm) [inline src/components/Tokens/TokenTable/TokenRow.tsx:437]: you can just do the uppercase in css

File: src/components/Tokens/TokenTable/TokenRow.tsx [hunk 6]
```diff
@@ -458,7 +457,7 @@ export default function LoadedRow({
   const exploreTokenSelectedEventProperties = {
     chain_id: filterNetwork,
     token_address: tokenAddress,
-    token_symbol: token?.symbol,
+    token_symbol: tokenData?.symbol,
```
Comment (vm) [inline src/components/Tokens/TokenTable/TokenRow.tsx:460]: then use tokenSymbol variable here

File: src/components/Tokens/TokenTable/TokenTable.tsx [hunk 7]
```diff
@@ -48,34 +47,29 @@ const TokenRowsContainer = styled.div`
...
+      shownTokens.filter((token) => {
         if (!filterString) {
           return true
         }
         const lowercaseFilterString = filterString.toLowerCase()
-        const addressIncludesFilterString = tokenAddress.toLowerCase().includes(lowercaseFilterString)
-        const nameIncludesFilterString = tokenName.toLowerCase().includes(lowercaseFilterString)
-        const symbolIncludesFilterString = tokenSymbol.toLowerCase().includes(lowercaseFilterString)
+        const addressIncludesFilterString = token.address.toLowerCase().includes(lowercaseFilterString)
+        const nameIncludesFilterString = token?.name?.toLowerCase().includes(lowercaseFilterString)
```
Comment (vm) [inline src/components/Tokens/TokenTable/TokenTable.tsx:64]: can token be undefined?

File: src/components/Tokens/TokenTable/TokenTable.tsx [hunk 8]
```diff
@@ -48,34 +47,29 @@ const TokenRowsContainer = styled.div`
...
+        const symbolIncludesFilterString = token?.symbol?.toLowerCase().includes(lowercaseFilterString)
```
Comment (vm) [inline src/components/Tokens/TokenTable/TokenTable.tsx:65]: same

File: src/components/Tokens/TokenTable/TokenTable.tsx (hunk 7 above)
Comment (vm) [reply src/components/Tokens/TokenTable/TokenTable.tsx:64]: and will this work as expected when undefined?

File: src/components/Tokens/TokenTable/TokenTable.tsx [hunk 9]
```diff
@@ -94,40 +88,39 @@ function useSortedTokens(addresses: string[], tokenData: Record<string, TokenDat
...
         let b: number
         switch (sortCategory) {
           case Category.marketCap:
-            a = token1.marketCap
-            b = token2.marketCap
-            break
-          case Category.percentChange:
-            a = token1.delta
-            b = token2.delta
+            a = token1.marketCap?.value ?? 0
```
Comment (vm) [inline src/components/Tokens/TokenTable/TokenTable.tsx:104]: not convinced defaulting to 0 is correct here?

File: src/components/Tokens/TokenTable/TokenTable.tsx [hunk 10]
```diff
@@ -156,9 +149,8 @@ function LoadingTokenTable() {
 export default function TokenTable({ data, error, loading }: UseTopTokensResult) {
   const showFavorites = useAtomValue<boolean>(showFavoritesAtom)
   const timePeriod = useAtomValue<TimePeriod>(filterTimeAtom)
-  const topTokenAddresses = data ? Object.keys(data) : []
-  const filteredTokens = useFilteredTokens(topTokenAddresses)
-  const filteredAndSortedTokens = useSortedTokens(filteredTokens, data)
+  const filteredTokens = useFilteredTokens(data ?? [])
+  const filteredAndSortedTokens = useSortedTokens(filteredTokens)
```
Comment (vm) [inline src/components/Tokens/TokenTable/TokenTable.tsx:153]: sortedFilteredTokens

File: src/components/Tokens/TokenTable/TokenTable.tsx [hunk 11]
```diff
@@ -156,9 +149,8 @@ function LoadingTokenTable() {
 export default function TokenTable({ data, error, loading }: UseTopTokensResult) {
   const showFavorites = useAtomValue<boolean>(showFavoritesAtom)
   const timePeriod = useAtomValue<TimePeriod>(filterTimeAtom)
-  const topTokenAddresses = data ? Object.keys(data) : []
-  const filteredTokens = useFilteredTokens(topTokenAddresses)
-  const filteredAndSortedTokens = useSortedTokens(filteredTokens, data)
+  const filteredTokens = useFilteredTokens(data ?? [])
```
Comment (vm) [inline src/components/Tokens/TokenTable/TokenTable.tsx:152]: i'd allow useFilteredTokens to handle taking undefined

File: src/components/Tokens/TokenTable/TokenTable.tsx (hunk 9 above)
Comment (kayleegeorge) [reply src/components/Tokens/TokenTable/TokenTable.tsx:104]: ya same, if it's not there, should we just put it at the bottom? this is for sort right

File: src/graphql/data/TopTokenQuery.ts (hunk 3 above)
Comment (kayleegeorge) [reply src/graphql/data/TopTokenQuery.ts:20]: that's the type the api returns

File: src/components/Tokens/TokenTable/TokenTable.tsx (hunk 9 above)
Comment (vm) [reply src/components/Tokens/TokenTable/TokenTable.tsx:104]: yes

File: src/graphql/data/TopTokenQuery.ts (hunk 3 above)
Comment (vm) [reply src/graphql/data/TopTokenQuery.ts:20]: it returns `string | null | undefined`? what's the diff between when it returns null vs undefined?

File: src/components/Tokens/TokenTable/TokenTable.tsx [hunk 12]
```diff
@@ -48,34 +47,30 @@ const TokenRowsContainer = styled.div`
   width: 100%;
 `
 
-function useFilteredTokens(addresses: string[]) {
+function useFilteredTokens(tokens: TokenData[] | null | undefined) {
```
Comment (vm) [inline src/components/Tokens/TokenTable/TokenTable.tsx:50]: why null or undefined? it should be only one of these

File: src/graphql/data/TopTokenQuery.ts (hunk 3 above)
Comment (vm) [reply src/graphql/data/TopTokenQuery.ts:20]: i dont think any of these will be undefined t'll just be null

Comment (kayleegeorge) [reply src/graphql/data/TopTokenQuery.ts:20]: gives error when don't include undefined for those particular fields that still have it as an option O.o

**Review by vm — APPROVED** [review]

**Pull request merged**